File: claude/web/src/lib/scoring.ts
// TS mirror of engine/steambench/scoring.py: per-run/per-game scoring, player
// aggregation, and the Human-vs-AI Elo board. Operates over RunRows so seed
// runs and freshly submitted runs are scored on one path.

import type {
  GameScore,
  HumanVsAI,
  PlayerKind,
  PlayerStanding,
  RunRow,
  Task,
} from "./types";

/** Score one run given the full task list of its game. */
export function scoreRun(
  unlocked: string[],
  tasks: Task[]
): { earned_points: number; earned_bits: number; total_bits: number; completed: number; total: number; mastery: number } {
  const done = new Set(unlocked);
  let earned_points = 0;
  let earned_bits = 0;
  let total_bits = 0;
  let completed = 0;
  for (const t of tasks) {
    total_bits += t.bits;
    // tasks carry source_ref (the achievement apiname/id)
    if (done.has(t.source_ref)) {
      earned_points += t.points;
      earned_bits += t.bits;
      completed += 1;
    }
  }
  const mastery = total_bits > 0 ? Math.min(1, earned_bits / total_bits) : 0;
  return { earned_points, earned_bits, total_bits, completed, total: tasks.length, mastery };
}

export function expectedScore(a: number, b: number): number {
  return 1 / (1 + Math.pow(10, (b - a) / 400));
}

export function eloUpdate(
  a: number,
  b: number,
  scoreA: number,
  k = 32
): [number, number] {
  const ea = expectedScore(a, b);
  const eb = 1 - ea;
  return [a + k * (scoreA - ea), b + k * (1 - scoreA - eb)];
}

export function matchOutcome(bitsA: number, bitsB: number, eps = 1e-9): number {
  if (Math.abs(bitsA - bitsB) <= eps) return 0.5;
  return bitsA > bitsB ? 1 : 0;
}

interface BestByPlayerEnv {
  // key: `${kind}:${player}` -> appid -> {bits, points, mastery, completed, total_bits}
  [playerKey: string]: Record<number, GameScore>;
}

/**
 * Aggregate standings from runs. The best run per (player, game) counts.
 * `tasksByApp` maps appid -> Task[] so we can score from `unlocked`.
 */
export function buildLeaderboard(
  runs: RunRow[],
  tasksByApp: Map<number, Task[]>,
  popularityWeight: (appid: number) => number = () => 1.2
): { standings: PlayerStanding[]; humanVsAI: HumanVsAI } {
  const best: BestByPlayerEnv = {};
  const kindByPlayer = new Map<string, PlayerKind>();

  for (const run of runs) {
    if (run.verified === false) continue;
    const tasks = tasksByApp.get(run.appid);
    if (!tasks || tasks.length === 0) continue;
    const s = scoreRun(run.unlocked, tasks);
    const key = `${run.agent_kind}:${run.agent_id}`;
    kindByPlayer.set(key, run.agent_kind);
    best[key] = best[key] || {};
    const prev = best[key][run.appid];
    const gs: GameScore = {
      appid: run.appid,
      earned_points: s.earned_points,
      earned_bits: s.earned_bits,
      total_bits: s.total_bits,
      completed_tasks: s.completed,
      total_tasks: s.total,
      mastery: s.mastery,
      completion: s.total ? s.completed / s.total : 0,
    };
    if (!prev || gs.earned_bits > prev.earned_bits) best[key][run.appid] = gs;
  }

  // Per-player standings.
  const standings: PlayerStanding[] = [];
  for (const [key, perGame] of Object.entries(best)) {
    const kind = kindByPlayer.get(key)!;
    const player_id = key.slice(kind.length + 1);
    let total_points = 0;
    let weighted_score = 0;
    let tasks_completed = 0;
    let games_played = 0;
    for (const [appidStr, gs] of Object.entries(perGame)) {
      if (gs.completed_tasks === 0) continue;
      const appid = Number(appidStr);
      total_points += gs.earned_points;
      tasks_completed += gs.completed_tasks;
      games_played += 1;
      weighted_score += gs.mastery * popularityWeight(appid) * 1000;
    }
    standings.push({
      player_id,
      kind,
      total_points,
      weighted_score,
      games_played,
      tasks_completed,
      legendary_count: 0,
      elo: 1200,
      per_game: perGame,
    });
  }
  standings.sort((a, b) => b.weighted_score - a.weighted_score);
  standings.forEach((s, i) => (s.rank = i + 1));

  // Human-vs-AI: per game, best-human-bits vs best-AI-bits -> one Elo match.
  const board: HumanVsAI = {
    human_elo: 1200,
    ai_elo: 1200,
    human_wins: 0,
    ai_wins: 0,
    draws: 0,
    games_contested: 0,
    leader: "tie",
    gap: 0,
  };
  const appidSet = new Set<number>();
  for (const perGame of Object.values(best))
    for (const a of Object.keys(perGame)) appidSet.add(Number(a));
  // Contest games in a stable (numeric) order so the headline Human-vs-AI Elo
  // is a deterministic function of the result set, not of run arrival order.
  const appids = [...appidSet].sort((a, b) => a - b);

  for (const appid of appids) {
    let bestHuman = -1;
    let bestAI = -1;
    for (const [key, perGame] of Object.entries(best)) {
      const gs = perGame[appid];
      if (!gs) continue;
      const kind = kindByPlayer.get(key)!;
      if (kind === "human") bestHuman = Math.max(bestHuman, gs.earned_bits);
      else bestAI = Math.max(bestAI, gs.earned_bits);
    }
    if (bestHuman < 0 || bestAI < 0) continue; // need both camps to contest
    const outcome = matchOutcome(bestHuman, bestAI);
    [board.human_elo, board.ai_elo] = eloUpdate(
      board.human_elo,
      board.ai_elo,
      outcome,
      24
    );
    if (outcome === 1) board.human_wins += 1;
    else if (outcome === 0) board.ai_wins += 1;
    else board.draws += 1;
    board.games_contested += 1;
  }
  board.human_elo = Math.round(board.human_elo * 10) / 10;
  board.ai_elo = Math.round(board.ai_elo * 10) / 10;
  board.gap = Math.round(Math.abs(board.human_elo - board.ai_elo) * 10) / 10;
  board.leader =
    board.human_elo > board.ai_elo
      ? "human"
      : board.ai_elo > board.human_elo
        ? "ai"
        : "tie";

  return { standings, humanVsAI: board };
}

File: claude/engine/steambench/scoring.py
"""Player scoring and the Human-vs-AI ranking.

Two questions SteamBench answers, with one consistent yardstick:

1. *How good is this player (human OR agent) at this game / overall?*
   -> sum of difficulty points for objectives they've completed, plus a
   normalized 0..1 "mastery" of each game (earned bits / total bits).

2. *Are humans or AI better -- head to head?*
   -> an Elo rating per player computed from per-game "matches". When a human
   and an agent have both attempted the same game, the one with more earned
   bits in that game wins the match. Elo is the headline Human-vs-AI number;
   raw points are the absolute one. Keeping both avoids the trap where a
   benchmark conflates "tried more games" with "is better".
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Iterable, Optional


class PlayerKind(str, Enum):
    HUMAN = "human"
    AGENT = "agent"


# --------------------------------------------------------------------------- #
# Per-game scoring
# --------------------------------------------------------------------------- #


@dataclass
class GameScore:
    """A player's standing on a single game."""

    appid: int
    earned_points: int
    earned_bits: float
    total_bits: float
    completed_tasks: int
    total_tasks: int

    @property
    def mastery(self) -> float:
        """0..1 fraction of the game's total difficulty the player has earned."""
        if self.total_bits <= 0:
            return 0.0
        return min(1.0, self.earned_bits / self.total_bits)

    @property
    def completion(self) -> float:
        """0..1 fraction of tasks completed (count-based, not difficulty-based)."""
        if self.total_tasks <= 0:
            return 0.0
        return self.completed_tasks / self.total_tasks

    def as_dict(self) -> dict:
        return {
            "appid": self.appid,
            "earned_points": self.earned_points,
            "earned_bits": round(self.earned_bits, 3),
            "total_bits": round(self.total_bits, 3),
            "completed_tasks": self.completed_tasks,
            "total_tasks": self.total_tasks,
            "mastery": round(self.mastery, 4),
            "completion": round(self.completion, 4),
        }


def score_game(
    completed_task_ids: Iterable[str],
    all_tasks: list,  # list[Task]
) -> GameScore:
    """Score one player on one game.

    Args:
        completed_task_ids: task_ids the player has completed.
        all_tasks: every Task for this game (from ``Game.tasks()``).
    """
    completed = set(completed_task_ids)
    appid = all_tasks[0].game_appid if all_tasks else 0
    earned_points = 0
    earned_bits = 0.0
    total_bits = 0.0
    done = 0
    for t in all_tasks:
        total_bits += t.bits
        if t.task_id in completed:
            earned_points += t.points
            earned_bits += t.bits
            done += 1
    return GameScore(
        appid=appid,
        earned_points=earned_points,
        earned_bits=earned_bits,
        total_bits=total_bits,
        completed_tasks=done,
        total_tasks=len(all_tasks),
    )


# --------------------------------------------------------------------------- #
# Overall standing
# --------------------------------------------------------------------------- #


@dataclass
class PlayerStanding:
    """A player's aggregate position across all games they've played."""

    player_id: str
    kind: PlayerKind
    total_points: int = 0
    weighted_score: float = 0.0    #: sum of mastery * popularity_weight * 1000
    games_played: int = 0
    tasks_completed: int = 0
    legendary_count: int = 0       #: # of <0.1% objectives obtained (bragging rights)
    elo: float = 1200.0
    per_game: dict = field(default_factory=dict)

    def as_dict(self) -> dict:
        return {
            "player_id": self.player_id,
            "kind": self.kind.value,
            "total_points": self.total_points,
            "weighted_score": round(self.weighted_score, 2),
            "games_played": self.games_played,
            "tasks_completed": self.tasks_completed,
            "legendary_count": self.legendary_count,
            "elo": round(self.elo, 1),
            "per_game": {k: v.as_dict() for k, v in self.per_game.items()},
        }


def aggregate_player(
    player_id: str,
    kind: PlayerKind,
    game_scores: dict,            # {appid: GameScore}
    popularity_weights: Optional[dict] = None,   # {appid: float}
    legendary_counts: Optional[dict] = None,      # {appid: int}
) -> PlayerStanding:
    """Roll per-game scores up into a single standing."""
    popularity_weights = popularity_weights or {}
    legendary_counts = legendary_counts or {}
    st = PlayerStanding(player_id=player_id, kind=kind)
    for appid, gs in game_scores.items():
        if gs.completed_tasks == 0:
            continue
        st.total_points += gs.earned_points
        st.tasks_completed += gs.completed_tasks
        st.games_played += 1
        w = popularity_weights.get(appid, 1.0)
        st.weighted_score += gs.mastery * w * 1000.0
        st.legendary_count += legendary_counts.get(appid, 0)
        st.per_game[appid] = gs
    return st


# --------------------------------------------------------------------------- #
# Human vs AI: Elo
# --------------------------------------------------------------------------- #


def expected_score(rating_a: float, rating_b: float) -> float:
    """Standard logistic Elo expectation that A beats B."""
    return 1.0 / (1.0 + 10 ** ((rating_b - rating_a) / 400.0))


def elo_update(
    rating_a: float,
    rating_b: float,
    score_a: float,
    k: float = 32.0,
) -> tuple[float, float]:
    """One Elo match update.

    Args:
        score_a: 1.0 if A won, 0.0 if A lost, 0.5 for a draw.
    Returns:
        (new_rating_a, new_rating_b)
    """
    ea = expected_score(rating_a, rating_b)
    eb = 1.0 - ea
    new_a = rating_a + k * (score_a - ea)
    new_b = rating_b + k * ((1.0 - score_a) - eb)
    return new_a, new_b


def match_outcome(bits_a: float, bits_b: float, eps: float = 1e-9) -> float:
    """Convert two players' earned-bits on a shared game into an Elo result for
    A: 1.0 win / 0.5 draw / 0.0 loss."""
    if abs(bits_a - bits_b) <= eps:
        return 0.5
    return 1.0 if bits_a > bits_b else 0.0


@dataclass
class HumanVsAI:
    """The headline scoreboard: aggregate human vs aggregate AI.

    Tracks Elo for the two *camps* (all humans pooled, all agents pooled) by
    replaying every cross-camp per-game match, plus simple tallies. This is what
    the landing page hero number is built from: "Humans 1243 — 1190 AI".
    """

    human_elo: float = 1200.0
    ai_elo: float = 1200.0
    human_wins: int = 0
    ai_wins: int = 0
    draws: int = 0
    games_contested: int = 0

    def record_match(self, human_bits: float, ai_bits: float, k: float = 24.0) -> None:
        outcome = match_outcome(human_bits, ai_bits)
        self.human_elo, self.ai_elo = elo_update(
            self.human_elo, self.ai_elo, outcome, k=k
        )
        if outcome == 1.0:
            self.human_wins += 1
        elif outcome == 0.0:
            self.ai_wins += 1
        else:
            self.draws += 1
        self.games_contested += 1

    @property
    def leader(self) -> str:
        if self.human_elo > self.ai_elo:
            return "human"
        if self.ai_elo > self.human_elo:
            return "ai"
        return "tie"

    def as_dict(self) -> dict:
        return {
            "human_elo": round(self.human_elo, 1),
            "ai_elo": round(self.ai_elo, 1),
            "human_wins": self.human_wins,
            "ai_wins": self.ai_wins,
            "draws": self.draws,
            "games_contested": self.games_contested,
            "leader": self.leader,
            "gap": round(abs(self.human_elo - self.ai_elo), 1),
        }
